WalletWasabi/WabiSabi/Client/CredentialDependencies/RequestNode.cs

```diff
@@ -0,0 +1,29 @@
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using WalletWasabi.Helpers;
+
+namespace WalletWasabi.WabiSabi.Client.CredentialDependencies
+{
+	// make private inner class of Graph?
+	public class RequestNode
+	{
+		public RequestNode(IEnumerable<long> values, int inDegree, int outDegree, int zeroOnlyOutDegree)
+		{
+			Values = Guard.InRange(nameof(values), values, DependencyGraph.K, DependencyGraph.K).ToImmutableArray();
+			MaxInDegree = inDegree;
+			MaxOutDegree = outDegree;
+			MaxZeroOnlyOutDegree = zeroOnlyOutDegree;
+		}
+
+		public ImmutableArray<long> Values { get; }
+
+		// TODO ImmutableArray<int> (uint?)
```

Comment:
long is fine IMO, uint just causing casts everywhere (usually).

Reply:
This comment is about making the `MaxInDegree`, MaxOutDegree` and `MaxZeroOnlyOutDegree` parameterized per type, sorry it's very vaguee... Since the degree is a natural number, not an integer uint might make sense but yeah, the fact that C# uses int for size types makes that noisy because of all the casting.

The reason we might need to do this is to represent given credentials, which need to be a node with a max out degree of 1 but only for a single credential type, and a max degree of 0 everywhere else.

Suppose we have an error in one of the connection confirmations that leaves the client with one input missing. It would need to decompose the actual balance, and then compute a new graph, but it already has some credentials from the inputs it did manage to confirm. Maybe there are other possible examples, but that's the only one that I think of that is actually of any importance right now.

I think this comment can be deleted. For now I haven't implemented this because it's not directly needed, and i think in general avoiding failing rounds is more of a priority for mainnet release than an immediate goal.